hfc/fabric/client.py

```diff
@@ -1619,17 +1623,57 @@ def cc_event(self, cc_event, block_number, tx_id, tx_status):
 
         # send proposal
-        responses, proposal, header = channel.send_tx_proposal(tx_context,
-                                                               target_peers)
+        responses, proposal, header = channel.send_tx_proposal(tx_context, target_peers)
 
         # The proposal return does not contain the transient map
         # because we do not sent it in the real transaction later
-        res = await asyncio.gather(*responses)
+        res = await asyncio.gather(*responses, return_exceptions=True)
+        failed_res = list(map(lambda x: isinstance(x, _MultiThreadedRendezvous), res))
 
-        # if proposal was not good, return
-        if not all([x.response.status == 200 for x in res]):
-            return res[0].response.message
+        # remove failed_res from res, orderer will take care of unmet policy (can be different between app,
+        # you should costumize this method to your own needs)
+        if any(failed_res):
+            res = list(filter(lambda x: hasattr(x, 'response') and x.response.status == 200, res))
+
+            # should we retry on failed?
+            if grpc_broker_unavailable_retry:
+                _logger.debug(f'Retry on failed proposal responses')
+
+                retry = 0
+
+                # get failed peers
+                failed_target_peers = list(itertools.compress(target_peers, failed_res))
+
+                while retry < grpc_broker_unavailable_retry:
+                    _logger.debug(f'Retrying getting proposal responses from peers:'
+                                  f' {[x.name for x in failed_target_peers]}, retry: {retry}')
+
+                    retry_responses, _, _ = channel.send_tx_proposal(tx_context, failed_target_peers)
+                    retry_res = await asyncio.gather(*retry_responses, return_exceptions=True)
+
+                    # get failed res
+                    failed_res = list(map(lambda x: isinstance(x, _MultiThreadedRendezvous), retry_res))
+
+                    # add successful responses to res and recompute failed_target_peers
+                    res += list(filter(lambda x: hasattr(x, 'response') and x.response.status == 200, retry_res))
+                    failed_target_peers = list(itertools.compress(failed_target_peers, failed_res))
+
+                    if len(failed_target_peers) == 0:
+                        break
+
+                    retry += 1
+                    # TODO should we use a backoff?
+                    _logger.debug(f'Retry in {grpc_broker_unavailable_retry_delay}ms')
+                    sleep(grpc_broker_unavailable_retry_delay / 1000)  # milliseconds
+
+                if len(failed_target_peers) > 0:
+                    # TODO should we really raise or send to the orderer without issue?
+                    raise Exception(f'Could not reach peer grpc broker {[x.name for x in failed_target_peers]}'
+                                    f' even after {grpc_broker_unavailable_retry} retries.')
```

Comment:
Can you put a option to choose if we want to send or not the proposal even if endorsement with some brokers have failed ?